src/drt/src/dr/FlexDR.cpp

```diff
@@ -561,57 +561,642 @@ void FlexDR::getBatchInfo(int& batchStepX, int& batchStepY)
+ *
+ * @param drv_boxes A list of gcell rectangles(with the gcell indices) that hold
+ * the violations.
+ * @returns a list of merged boxes.
+ */
+std::vector<Rect> mergeBoxes(std::vector<Rect>& drv_boxes)
+{
+  std::vector<Rect> merge_boxes;
+  for (auto& box : drv_boxes) {
+    merge_boxes.emplace_back(box);
+  }
+  bool merged;
+  do {
+    merged = false;
+    for (auto it1 = merge_boxes.begin(); it1 != merge_boxes.end(); ++it1) {
+      for (auto it2 = it1 + 1; it2 != merge_boxes.end(); ++it2) {
+        Rect merge_box = (*it1);
+        merge_box.merge((*it2));
+        if (merge_box.dx() > 4 || merge_box.dy() > 4) {
+          continue;
+        }
+        (*it1).merge((*it2));
+        merge_boxes.erase(it2);
+        merged = true;
+        break;
+      }
+      if (merged) {
+        break;
+      }
+    }
+  } while (merged);
+  return merge_boxes;
+}
+/**
+ * @brief Checks if the passed rectangle intersects with any other rectangle in
+ * the grid.
+ *
+ * @param grid A 2-d grid representing the current grid. (-1 value means
+ * unoccupied).
+ * @param rect_id the current rectangle id to disregard in the check.
+ * @return True if the rectangle intersects with an occupied cell by another
+ * rectangle.
+ */
+bool hasOtherRect(const std::vector<std::vector<int>>& grid,
+                  const Rect& rect,
+                  const int rect_id)
+{
+  return std::any_of(grid.begin() + rect.xMin(),
+                     grid.begin() + rect.xMax() + 1,
+                     [&](const std::vector<int>& row) {
+                       return std::any_of(row.begin() + rect.yMin(),
+                                          row.begin() + rect.yMax() + 1,
+                                          [rect_id](int id) {
+                                            return id != -1 && id != rect_id;
+                                          });
+                     });
+}
+/**
+ * @brief Expands the passed rectangle in the required direction.
+ *
+ * The function expands the passed rectangle in the required direction a gcell
+ * at a time with the following conditions:
+ * - The expansion is less than or equal to the max_expansion
+ * - The expansion does not result a rectangle that intersects with another
+ *   rectangle in the grid.
+ * @param box The rectangle to be expanded.
+ * @param id The rectangle's id.
+ * @param grid A 2-d grid representing the current grid. (-1 value means
+ * unoccupied).
+ * @param dir The direction of the expansion (East, West, North, South).
+ * @return The number of expanded gcells (less than or equal to max_expansion)
+ */
+int expandBox(Rect& box,
+              const int id,
+              const std::vector<std::vector<int>>& grid,
+              const frDirEnum dir,
+              const int max_expansion)
+{
+  auto min_idx = box.ll();
+  auto max_idx = box.ur();
+  for (int i = 1; i <= max_expansion; i++) {
+    switch (dir) {
+      case frDirEnum::E:
+        max_idx.addX(1);
```

Comment:
Yes, because -1 index would cause an error in hasOtherRect function. The grid can hold more gcell indices than there actually is. So we can expand west and north as much as we want even if there are no gcells beyond. They are clipped later when we transform the gcell indices to real coordinates.